### PR TITLE
(PC-7021) App native - Favoris - Supprimer un favori

### PR DESCRIPTION
**Pull request by kopax-polyconseil**
Link to JIRA ticket: https://passculture.atlassian.net/browse/PC-7021

Also fix :

- https://passculture.atlassian.net/browse/PC-7504
- https://passculture.atlassian.net/browse/PC-7175

## Checklist

I have:

- [x] Made sure the title of my PR follows the convention `[PC-XXX] <summary>`.
- [x] Made sure my feature is working on the relevant real / virtual devices.
- [x] Written **unit tests** for my feature.
- [x] Added new reusable components to the AppComponents page and communicate to the team on slack
- [ ] Added a **screenshot** for UI tickets.
- [ ] Attached a **ticket number** for any added TODO/FIXME \
       (for tech tasks, give the best context about the TODO resolution: what/who/when).

## Screenshots

| \*iOS - [Phone name] | \*Android - [Phone name] | Tablet - [Phone name] | Little - [Phone name] |
| -------------------- | :----------------------: | --------------------: | --------------------: |
|                      |                          |                       |                       |

## Deploy hard

If native code (ios/android) was modified, **after** the PR is merged, on the master branch, upgrade the **app version** (+1 patch):

- if you want an hard deployment of the testing environment, use `yarn trigger:testing:deploy`

### CODE REVIEW

File: src/features/favorites/pages/useFavorites.ts [hunk 1]
```diff
@@ -1,28 +1,143 @@
...
   const { isLoggedIn } = useAuthContext()
 
-  return useQuery<PaginatedFavoritesResponse>('favorites', () => api.getnativev1mefavorites(), {
+  return useQuery<PaginatedFavoritesResponse>(QUERY_KEY, () => api.getnativev1mefavorites(), {
     enabled: isLoggedIn,
   })
 }
 
-export function useIsFavorite(offerId: number) {
+export function useFromFavorite({ offerId, id }: { offerId?: number; id?: number }) {
```
[the remark this answers is not in the document]
Comment (BenStirrup) [reply src/features/favorites/pages/useFavorites.ts:127]: Plutôt `useFavorite` non ?

File: src/features/favorites/pages/useFavorites.ts [hunk 2]
```diff
@@ -1,28 +1,143 @@
...
+      }
+      if (onSuccess) {
+        onSuccess(data)
+      }
+    },
+    onMutate: ({ offerId }) => {
+      const previousFavorites = queryClient.getQueryData<PaginatedFavoritesResponse>(QUERY_KEY)
+      if (previousFavorites) {
+        queryClient.setQueryData(QUERY_KEY, {
+          ...previousFavorites,
```
Comment (BenStirrup) [inline src/features/favorites/pages/useFavorites.ts:94]: Ce spreading est inutile non ?

File: src/features/favorites/pages/useFavorites.ts [hunk 3]
```diff
@@ -1,28 +1,143 @@
...
+export function useAddFavorite({ onSuccess, onError, onMutate, onSettled }: AddFavorite) {
+  const queryClient = useQueryClient()
   return useMutation((body: FavoriteRequest) => api.postnativev1mefavorites(body), {
-    onSuccess,
-    onError,
+    onSuccess: (data: FavoriteResponse) => {
+      const previousFavorites = queryClient.getQueryData<PaginatedFavoritesResponse>(QUERY_KEY)
+      if (previousFavorites) {
+        queryClient.setQueryData(QUERY_KEY, {
+          ...previousFavorites,
```
Comment (BenStirrup) [inline src/features/favorites/pages/useFavorites.ts:77]: Spreading inutile ?

File: src/features/favorites/pages/useFavorites.ts [hunk 4]
```diff
@@ -1,28 +1,143 @@
...
+      onSuccess,
+      onMutate: async (favoriteId) => {
+        await queryClient.cancelQueries(QUERY_KEY)
+        // Snapshot the previous value
+        const previousFavorites = queryClient.getQueryData<PaginatedFavoritesResponse>(QUERY_KEY)
+
+        // Optimistically update to the new value
+        if (favoriteId && previousFavorites) {
+          queryClient.setQueryData<PaginatedFavoritesResponse>(QUERY_KEY, {
+            ...previousFavorites,
```
Comment (BenStirrup) [inline src/features/favorites/pages/useFavorites.ts:47]: Spreading inutiles ?

File: src/features/favorites/pages/useFavorites.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (kopax-polyconseil) [reply src/features/favorites/pages/useFavorites.ts:127]: OK

File: src/features/offer/components/__tests__/OfferHeader.test.tsx [hunk 5]
```diff
@@ -146,20 +300,44 @@ const humanizedOfferId = 'AHD3A'
...
+  { isLoggedIn }: { isLoggedIn: boolean },
+  id: number = offerId,
+  {
+    hasAddFavoriteError,
+    hasRemoveFavoriteError,
+  }: { hasAddFavoriteError?: boolean; hasRemoveFavoriteError?: boolean } = {
+    hasAddFavoriteError: false,
+    hasRemoveFavoriteError: false,
+  }
+) {
```
[the remark this answers is not in the document]
Comment (kopax-polyconseil) [reply src/features/offer/components/__tests__/OfferHeader.test.tsx:313]: OK

File: src/features/offer/components/__tests__/OfferHeader.test.tsx [hunk 6]
```diff
@@ -115,21 +132,158 @@ describe('<OfferHeader />', () => {
...
+      fireEvent.press(getByTestId('icon-favorite'))
+      await waitForExpect(() => {
+        const mutateDataBetween = queryCache.find('favorites')?.state
+          ?.data as PaginatedFavoritesResponse
+        expect(
+          mutateDataBetween.favorites?.find(
+            (f: FavoriteResponse) => f.offer.id === addFavoriteJsonResponseSnap.offer.id
+          )?.offer.id
+        ).toBe(10000)
+        expect(queryByTestId('icon-favorite-filled')).toBeTruthy()
```
[the remark this answers is not in the document]
Comment (kopax-polyconseil) [reply src/features/offer/components/__tests__/OfferHeader.test.tsx:198]: OK

[the remark this answers is not in the document]
Comment (kopax-polyconseil) [reply src/features/offer/components/__tests__/OfferHeader.test.tsx:198]: (Mais je crois avoir fait ça pour cacher certains warning)

File: src/features/favorites/pages/useFavorites.ts (hunk 4 above)
Comment (kopax-polyconseil) [reply src/features/favorites/pages/useFavorites.ts:47]: Non, il y a `nbFavorites`, `pages` dedans, et ça doit être dans le prochain cache.

File: src/features/favorites/pages/useFavorites.ts (hunk 3 above)
Comment (kopax-polyconseil) [reply src/features/favorites/pages/useFavorites.ts:77]: Utile

File: src/features/favorites/pages/useFavorites.ts (hunk 2 above)
Comment (kopax-polyconseil) [reply src/features/favorites/pages/useFavorites.ts:94]: Utile

File: src/features/favorites/pages/useFavorites.ts [hunk 7]
```diff
@@ -1,28 +1,143 @@
...
+      }
+    },
+    onMutate: ({ offerId }) => {
+      const previousFavorites = queryClient.getQueryData<PaginatedFavoritesResponse>(QUERY_KEY)
+      if (previousFavorites) {
+        queryClient.setQueryData(QUERY_KEY, {
+          ...previousFavorites,
+          favorites: [
+            ...previousFavorites.favorites,
+            { id: Math.random(), offer: { id: offerId } },
```
[the remark this answers is not in the document]
Comment (BenStirrup) [reply src/features/favorites/pages/useFavorites.ts:97]: Pourquoi pas `id: undefined` alors ?

File: src/features/favorites/pages/useFavorites.ts (hunk 2 above)
Comment (BenStirrup) [reply src/features/favorites/pages/useFavorites.ts:94]: ok

File: src/features/favorites/pages/useFavorites.ts (hunk 3 above)
Comment (BenStirrup) [reply src/features/favorites/pages/useFavorites.ts:77]: ok

File: src/features/favorites/pages/useFavorites.ts (hunk 4 above)
Comment (BenStirrup) [reply src/features/favorites/pages/useFavorites.ts:47]: ok

**Review by BenStirrup — APPROVED** [review]

**Pull request merged**